### ISSUE TITLE
[DeepSeek] Potential memory leak, OOMs after a few steps

### ISSUE THREAD

**Issue by EugenHotaj** [pytorch/torchtitan#979]
### Bug description

I've been playing around with the [experimental DeepSeek code](https://github.com/pytorch/torchtitan/tree/main/torchtitan/experiments/deepseek_v3) and in-progress PRs and have a simple sft training loop implemented. The results on the first few steps look pretty promising and match an identical run we have with a different training framework.

However, I have not been able to train past a few steps because the job OOMs. All our training examples are the same sequence length so in theory if the first step succeeds the reset should as well. 

Instead, we observe that the memory grows after each training step. Below are some logs for a `DeepSeek-V2-Lite` run with `pp=2, ep=2` up until the job OOMs.

```
Creating model stage 0 of 2
Creating model stage 1 of 2
Creating model stage 0 of 2
Creating model stage 1 of 2
EP rank [0]: Created Symmetric Memory for MoE
EP rank [1]: Created Symmetric Memory for MoE
EP rank [0]: Created Symmetric Memory for MoE
EP rank [1]: Created Symmetric Memory for MoE
loss=1.515625
allocated: 31894.55 MB, reserved: 58676.00 MB
allocated: 32995.59 MB, reserved: 58678.00 MB
allocated: 23607.79 MB, reserved: 38456.00 MB
allocated: 22644.09 MB, reserved: 37930.00 MB
allocated: 48067.40 MB, reserved: 71704.00 MB
loss=1.9296875
allocated: 45621.36 MB, reserved: 65286.00 MB
allocated: 34495.37 MB, reserved: 49418.00 MB
allocated: 36627.83 MB, reserved: 51514.00 MB
allocated: 62918.86 MB, reserved: 84702.00 MB
loss=1.6328125
allocated: 59593.46 MB, reserved: 84682.00 MB
allocated: 49534.78 MB, reserved: 62938.00 MB
allocated: 46510.73 MB, reserved: 59692.00 MB
allocated: 78034.18 MB, reserved: 97744.00 MB
loss=1.875
allocated: 73243.11 MB, reserved: 94548.00 MB
allocated: 62594.60 MB, reserved: 75444.00 MB
allocated: 58295.73 MB, reserved: 70800.00 MB
allocated: 93120.65 MB, reserved: 114002.00 MB
loss=1.9296875
allocated: 86932.72 MB, reserved: 101184.00 MB
allocated: 75577.72 MB, reserved: 87698.00 MB
allocated: 70176.92 MB, reserved: 81938.00 MB
allocated: 107968.62 MB, reserved: 120578.00 MB
loss=1.703125
allocated: 100870.77 MB, reserved: 120570.00 MB
allocated: 88635.73 MB, reserved: 99480.00 MB
allocated: 82006.98 MB, reserved: 92282.00 MB
loss=1.734375
allocated: 122883.92 MB, reserved: 131216.00 MB
allocated: 114739.80 MB, reserved: 130336.00 MB
allocated: 93780.24 MB, reserved: 102500.00 MB
allocated: 101756.58 MB, reserved: 111374.00 MB
<OOM>
```

### Versions

Using pytorch nightlies.

**Comment by EugenHotaj** [comment]
Looks like the leak is happening somewhere in the ep<->dp shuffle kernels. If I comment out the `on_device_all_to_all_v()` calls inside `MOE.moe_forward` the memory stabilizes: 

```
Creating model stage 1 of 2
EP rank [1]: Created Symmetric Memory for MoE
EP rank [0]: Created Symmetric Memory for MoE
EP rank [0]: Created Symmetric Memory for MoE
EP rank [1]: Created Symmetric Memory for MoE
loss=18.0
allocated: 17888.48 MB, reserved: 69708.00 MB
allocated: 17888.48 MB, reserved: 69708.00 MB
allocated: 10532.33 MB, reserved: 44818.00 MB
allocated: 10532.33 MB, reserved: 44818.00 MB
allocated: 17909.48 MB, reserved: 72736.00 MB
loss=24.25
allocated: 17909.48 MB, reserved: 72736.00 MB
allocated: 10550.33 MB, reserved: 47284.00 MB
allocated: 10550.33 MB, reserved: 47284.00 MB
allocated: 17930.48 MB, reserved: 72758.00 MB
loss=15.875
allocated: 17930.48 MB, reserved: 72758.00 MB
allocated: 10568.33 MB, reserved: 47302.00 MB
allocated: 10568.33 MB, reserved: 47302.00 MB
allocated: 17951.48 MB, reserved: 72778.00 MB
loss=13.625
allocated: 17951.48 MB, reserved: 72778.00 MB
allocated: 10586.33 MB, reserved: 47320.00 MB
allocated: 10586.33 MB, reserved: 47320.00 MB
allocated: 17972.48 MB, reserved: 72800.00 MB
loss=12.875
allocated: 17972.48 MB, reserved: 72800.00 MB
allocated: 10604.33 MB, reserved: 47338.00 MB
allocated: 10604.33 MB, reserved: 47338.00 MB
allocated: 17993.48 MB, reserved: 72820.00 MB
loss=11.5
allocated: 17993.48 MB, reserved: 72820.00 MB
allocated: 10622.33 MB, reserved: 47356.00 MB
allocated: 10622.33 MB, reserved: 47356.00 MB
allocated: 18014.48 MB, reserved: 72840.00 MB
loss=11.5
allocated: 18014.48 MB, reserved: 72840.00 MB
allocated: 10640.33 MB, reserved: 47374.00 MB
allocated: 10640.33 MB, reserved: 47374.00 MB
allocated: 18035.48 MB, reserved: 72862.00 MB
loss=nan
allocated: 18035.48 MB, reserved: 72862.00 MB
allocated: 10658.33 MB, reserved: 47392.00 MB
allocated: 10658.33 MB, reserved: 47392.00 MB
allocated: 18056.48 MB, reserved: 72884.00 MB
```

**Comment by EugenHotaj** [comment]
I think this was because the bwd pass was never being called in the symmetric memory case.
